### PR TITLE
[inductor] Assume non-views are aligned, views are unaligned

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122159
* __->__ #122926

**TL;DR**: After this PR, if assume_aligned_inputs=False: Inductor will generate code based on un-guarded properties of the example input tensors; specifically, if an example input is a view, we generate unaligned triton code; and if it is not a view we generate aligned triton code (and clone any inputs that aren't aligned)

**Details**:
* If an input is a view, we generate unaligned triton. 
* If an input is not a view, we generate aligned triton and do a clone at runtime.
* This change is still hidden behind the `assume_aligned_inputs` flag
* This does not guard on whether tensors are views. I have a version that does guard here: https://github.com/pytorch/pytorch/pull/122867 but it has a failure in `test_issue106555` which is concerning; somehow, adding additional dynamo guards (and making no changes to inductor, because all the changes are still guarded in that PR) causes a size assertion failure in the backward pass.

**Alternatives**
* Guard on storage offset: this would be more accurate, because we would actually know whether the tensor is aligned. For example, there may be some views that are always aligned, but in this PR we will assume that they are not aligned. However, we ran into two issues with this: (1) dynamo guards took way too long, and (2) we'd start recompiling more than we'd like.
* Care about the users of the input: e.g. if it's a matmul, or maybe if it is loaded a lot of times, then it might be worthwhile to do the clone; otherwise, don't clone and generate unaligned code.
* Guard on tensor-is-a-view in case this property changes. As I mentioned above, I tried this in https://github.com/pytorch/pytorch/pull/122926 but ran into some issues; but I can take another look if we think this is important.
* We could also make assume_aligned_inputs=False by default (although we'd have to be careful about getting rid of the config)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/compile_fx.py [hunk 1]
```diff
@@ -799,17 +799,19 @@ def get_input_idxs_to_check(
     inputs: Union[List[torch.Tensor], Sequence[int]],
     static_input_idxs: Sequence[int],
 ) -> Sequence[int]:
-    def is_aligned(storage_offset, dtype):
+    def is_aligned(input):
+        # See Note: [Input Alignment handling in Inductor]
+        if input._is_view() and not config.assume_aligned_inputs:
+            return False
```
Comment (eellison) [inline torch/_inductor/compile_fx.py:805]: It's a little surprising to me that we're rechecking `_is_view()` here.. 

In our initial compilation we are determining which inputs we want to assume alignment of, and we take into account is_view() as a heuristic there. 

I would expect that we look up that list to check alignment instead of checking `is_view` here

Comment (davidberard98) [reply torch/_inductor/compile_fx.py:805]: Is your suggestion to look at the list from GraphLowering? I feel like that's a bit hard because there's a layer of code caching in between, and it deals with variables in the GraphLowering fx graph.

Would it be fine instead to put this logic in some util function that we check twice (once here, and once in GraphLowering)? Or is there additional reason that we really want to use one list and only perform the check once?
